File: UMN/CS5106 - Programming Languages/hw1/solution.pl
% A membership relation.
% Usage example: mem(1,[2,4,1,3])
mem(X,[X|_]).
mem(X,[Y|L]) :- mem(X,L).

% A subset relation.
% Usage example: subs([1,2],[3,2,4,1,5])
subs([],_).
subs([X|XS],L) :- mem(X,L), subs(XS,L).

% A prefix relation.
% Usage example: prefix([1,2],[1,2,3,4])
prefix([],_).
prefix([X|XS],[X|YS]):- prefix(XS,YS).

% A sublist relation
% Usage example: sublist([1,2],[3,4,1,2]).
sublist([],_).
sublist(XS,YS) :- prefix(XS,YS).
sublist(X,[_|YS]) :- sublist(X,YS).

File: UMN/CS5106 - Programming Languages/hw1/solution.ml
(* A prefix function.
   Usage example: prefix [1,2] [1,2,3,4] ; 
*)
fun prefix [] ys = true
 |  prefix xs [] = false
 |  prefix (x::xs) (y::ys) = if x = y then prefix xs ys else false

fun sublist [] ys = true
 |  sublist xs [] = false
 |  sublist (x::xs) (y::ys) = if x = y then if prefix xs ys then true else sublist (x::xs) ys else sublist (x::xs) ys
